bpaotu/bpaotu/query.py

```diff
@@ -179,6 +179,17 @@ def matching_sample_ids_and_environment(self):
         q = self._apply_filters(q, subq).order_by(SampleContext.id)
         return self._q_all_cached('matching_sample_ids_and_environment', q)
 
+    def matching_sample_headers(self, required_headers=None):
+        query_headers = [SampleContext.id, SampleContext.environment_id]
+
+        if required_headers:
+            for h in required_headers:
+                query_headers.append(getattr(SampleContext, h))
+
+        q = self._session.query(*query_headers)
+
+        return self._q_all_cached('matching_sample_headers', q)
```

Comment:
the cache bug is that you're not including required_headers in the cache key; so the key gets set for the first query done, and re-used even if the headers change.

We should sit down together and restructure how the caching code works. A python decorator that marshals function arguments sounds like a nice way. Let's look at this together.